### DISCUSSION TITLE
Apply a single filter expression to many fields

### DISCUSSION THREAD

**Discussion by NikosAlexandris** [johnkerl/miller#613] [Q&A]
The following filter expression works just fine (for #608).
```
filter 'is_numeric($ln) || is_numeric($le) || is_numeric($mn)'
```

What about automagically applying `is_numeric()` to all of the input fields?
And, would this be possible to limit only to fields after `cut`ting?

I am unsuccessfully trying something like
```
filter 'for (k,v in $*) { is_numeric(v) }'
```

**Comment by aborruso** [comment]
Does your goal is to filter every row in which at least one cell is a number?

Are all your fields numeric?

**Reply by NikosAlexandris** [reply]
> Does your goal is to filter every row in which at least one cell is a number?

Filter every record of selected fields/columns.

> Are all your fields numeric?

The ones I am interested to fill, yes.
I want to apply 'your' trick for selected columns. I have got my way to select the columns of interest. I want to avoid writing `is_numeric()` for every column I want. There are tenths or can be more of them.

**Comment by johnkerl** [comment]
@NikosAlexandris 

> I am unsuccessfully trying something like
> filter 'for (k,v in $*) { is_numeric(v) }'

Can you please include the full input and output of your unsuccessful attempt?

**Reply by NikosAlexandris** [reply]
> @NikosAlexandris
> 
> > I am unsuccessfully trying something like
> > filter 'for (k,v in $*) { is_numeric(v) }'
> 
> Can you please include the full input and output of your unsuccessful attempt?

I cannot reproduce the "faulty behaviour" any more 🤔.  It works now -- see answer below. I must have mistyped something... ? Though I just revisited my command line history.

**Reply by NikosAlexandris** [reply]
OK, I think I confused the initial unsuccessful attempt here with the supposedly working following (**now deleted**) examples. Will re-investigate.

**Reply by NikosAlexandris** [reply]
I still need to put some effort to learn loops. So, not the same example, but the important thing is to understand a bit `filter` and `for` looping first.

With 
```bash
➜  mlr.go cat test.csv
a1,a2,a3,b1,b2,b3,c1,c2,c3
0.5,0.2,0.3,0.3,0.4,0.5,0.6,0.7,0.8
```

how do I operate only on fields that start with 'a', for example?

**Reply by johnkerl** [reply]
@NikosAlexandris for the all-field-numeric check you need to "and" the `is_numeric` checks together. For languages with closures, like Ruby or JavaScript, this would be a use for `.every`. But in the Miller DSL you need to explictly "and":

```
$ cat mixed.csv
a,b,c
1,2,3
4,5,x
s,t,u
7,8,9
```

Then either

```
mlr --csv --from mixed.csv put '
  all_numeric = true;
  for (k,v in $*) {
    if (!is_numeric(v)) {
      all_numeric = false;
      break;
    }
  }
  filter(all_numeric);
'
```

or

```
mlr --csv --from mixed.csv filter '
  all_numeric = true;
  for (k,v in $*) {
    if (!is_numeric(v)) {
      all_numeric = false;
      break;
    }
  }
  all_numeric;
'
```

produces

```
a,b,c
1,2,3
7,8,9
```

**Reply by johnkerl** [reply]
For

> how do I operate only on fields that start with 'a', for example?

you can do this:

```
$ cat wide.csv
a1,a2,a3,b1,b2,b3,c1,c2,c3
5,2,3,3,4,5,6,7,8
a,b,c,3,4,5,6,7,8
5,2,3,3,4,5,x,y,z
```

then

```
mlr --csv --from wide.csv put '
  all_a_numeric = true;
  for (k,v in $*) {
    if (k =~ "^a.*$") {
      if (!is_numeric(v)) {
        all_a_numeric = false;
        break;
      }
    }
  }
  filter(all_a_numeric);
'
```

produces

```
a1,a2,a3,b1,b2,b3,c1,c2,c3
5,2,3,3,4,5,6,7,8
5,2,3,3,4,5,x,y,z
```

**Reply by NikosAlexandris** [reply]
@aborruso @johnkerl I think the right question is the one by @aborruso:

> Does your goal is to filter every row in which at least one cell is a number?

And my answer/description didn't help much. I should have demonstrated it better. I want to operate on fields/columns whose records are all numeric. So, column-wise. And not row-wise.

For example, with 
```bash
➜  cat mixed_2.csv
a,b,c
1-,2,3
4-,5,x
3-,4,u
7-,8,9
```
the following
```bash
➜  mlr.go --csv --from mixed_2.csv put '
  all_numeric = true;
  for (k,v in $*) {
    if (!is_numeric(v)) {
      all_numeric = false;
      break;
    }
  }
  filter(all_numeric);
'
```
won't give any output, whereas I am targetting at
```bash
➜  cat mixed_2.csv
b
2
5
4
8
```

**Reply by NikosAlexandris** [reply]
@aborruso correctly mentioned elsewhere to loop over each column. I just thought it's possible and worth learning how to do this with Miller only.

**Reply by NikosAlexandris** [reply]
Overall I think I would like to understand a bit more on how to work in a column-wise manner.

**Reply by johnkerl** [reply]
Ah OK @NikosAlexandris I get it! This will require https://johnkerl.org/miller6/operating-on-all-records/ as it's non-streaming (we need to see all rows to be sure `b` is an all-numeric column) -- we can make an example for sure. :)

**Reply by aborruso** [reply]
@NikosAlexandris it's not what you want. A stupid bash way could be the one below.

Starting from this csv input

```
a,b,c
1,2,a
3,5,t
4,7,o
```

I calculate the mean for each field. If I have `1` status, that field is not a number.

```bash
#!/bin/bash

set -x

if [ -f ./test ]; then
  rm ./test
fi

echo -e "$(mlr --c2n -N cat input.csv | head -n +1 | sed -r 's/ /\n/g')" | while read line; do
  test=$(mlr --csv stats1 -a mean -f "$line" input.csv)
  status=$?
  echo "field=$line,t=$status" >>./test
done

numericFields=$(mlr --ocsv filter '$t==0' then reshape -s field,t test | head -n +1)

mlr --csv cut -f $numericFields input.csv
```

The output is 

```
a,b
1,2
3,5
4,7
```

**Reply by johnkerl** [reply]
@NikosAlexandris 

```
$ cat mixed.csv
a,b,c
1,2,3
1,2,3
1,2,3
1,2,3
1,2,3
1,x,3
1,2,3
1,2,3
```

```
mlr --csv --from mixed.csv put -q '
  # Accumulate a list of all records
  @records[NR] = $*;
  # Accumulate a map of all field names seen
  for (k in $*) {
    @column_is_numeric[k] = true;
  }

  end {
    # See which column names are entirely numeric
    for (nr, record in @records) {
      for (k, v in record) {
        if (!is_numeric(v)) {
          @column_is_numeric[k] = false;
        }
      }
    }

    # Loop over retained records, unsetting the column
    # name if any record has non-numeric data in that column.
    for (nr, record in @records) {
      for (k, v in record) {
        if (!@column_is_numeric[k]) {
          unset record[k]
        }
      }
      emit record;
    }
  }
'
```

```
a,c
1,3
1,3
1,3
1,3
1,3
1,3
1,3
1,3
```
